kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/SharedInformerFactory.java

```diff
@@ -89,10 +86,134 @@ public SharedInformerFactory(ExecutorService threadPool, OkHttpClient okHttpClie
+   * @param resyncPeriodInMillis resync period in milliseconds
+   * @param <T> the type parameter (should extend {@link io.fabric8.kubernetes.api.model.HasMetadata} and implement {@link io.fabric8.kubernetes.api.model.Namespaced})
+   * @param <L> the type's list parameter (should extend {@link io.fabric8.kubernetes.api.model.KubernetesResourceList}
+   * @return the shared index informer
+   */
+  public synchronized <T extends HasMetadata, L extends KubernetesResourceList<T>> SharedIndexInformer<T> sharedIndexInformerForCustomResource(Class<T> apiTypeClass, Class<L> apiListTypeClass, OperationContext operationContext, long resyncPeriodInMillis) {
+    return sharedIndexInformerForCustomResource(CustomResourceDefinitionContext.fromCustomResourceType(apiTypeClass), apiTypeClass, apiListTypeClass, operationContext, resyncPeriodInMillis);
+  }
+
+  /**
+   * Constructs and returns a shared index informer with resync period specified for custom resources. You can use this
+   * method to specify namespace in {@link OperationContext} if you want to monitor for events in a dedicated namespace
+   * only or provide other filtering options.
+   *
+   * @param crdContext {@link CustomResourceDefinitionContext} Context for CustomResourceDefinition
+   * @param apiTypeClass apiType class
+   * @param resyncPeriodInMillis resync period in milliseconds
+   * @param <T> the type parameter (should extend {@link io.fabric8.kubernetes.api.model.HasMetadata} and implement {@link io.fabric8.kubernetes.api.model.Namespaced})
+   * @return the shared index informer
+   */
+  public synchronized <T extends HasMetadata> SharedIndexInformer<T> sharedIndexInformerForCustomResource(CustomResourceDefinitionContext crdContext, Class<T> apiTypeClass, long resyncPeriodInMillis) {
+    return sharedIndexInformerForCustomResource(crdContext, apiTypeClass, CustomResourceOperationsImpl.inferListType(apiTypeClass), resyncPeriodInMillis);
+  }
+
+  /**
+   * Constructs and returns a shared index informer with resync period specified for custom resources. You can use this
+   * method to specify namespace in {@link OperationContext} if you want to monitor for events in a dedicated namespace
+   * only or provide other filtering options.
+   *
+   * @param crdContext {@link CustomResourceDefinitionContext} Context for CustomResourceDefinition
+   * @param apiTypeClass apiType class
+   * @param operationContext operation context
+   * @param resyncPeriodInMillis resync period in milliseconds
+   * @param <T> the type parameter (should extend {@link io.fabric8.kubernetes.api.model.HasMetadata} and implement {@link io.fabric8.kubernetes.api.model.Namespaced})
+   * @return the shared index informer
+   */
+  public synchronized <T extends HasMetadata> SharedIndexInformer<T> sharedIndexInformerForCustomResource(CustomResourceDefinitionContext crdContext, Class<T> apiTypeClass, OperationContext operationContext, long resyncPeriodInMillis) {
```

Comment:
What should be done when the information provided by the CRDContext clashes with what could be derived from the api type class?

Reply:
As per our discussion, I'm removing this method in favor of new improved methods. Now CRDContext should be built from POJO annotations or opinionated defaults.